include/deal.II/lac/trilinos_tpetra_vector.templates.h

```diff
@@ -643,8 +750,45 @@ namespace LinearAlgebra
 
     template <typename Number>
     void
-    Vector<Number>::compress(const VectorOperation::values /*operation*/)
-    {}
+    Vector<Number>::compress(const VectorOperation::values operation,
+                             const IndexSet               &locally_owned)
+    {
+      if (!compressed)
+        {
+          Tpetra::CombineMode tpetra_operation = Tpetra::ZERO;
+          if (operation == VectorOperation::insert)
+            tpetra_operation = Tpetra::INSERT;
+          else if (operation == VectorOperation::add)
+            tpetra_operation = Tpetra::ADD;
+          else
+            Assert(false, ExcNotImplemented());
+
+          if (locally_owned != IndexSet())
+            {
+              locally_owned_map.reset();
+              locally_owned_map =
+                locally_owned.make_tpetra_map_rcp(mpi_comm(), false);
+            }
+
+          if (locally_relevant_map.is_null())
+            locally_relevant_map =
+              Teuchos::rcp_const_cast<MapType>(vector->getMap());
+
+          Assert(!locally_relevant_map.is_null(), ExcMissingIndexSet());
+          Assert(!locally_owned_map.is_null(), ExcMissingIndexSet());
+
+          VectorType dummy = VectorType(locally_owned_map.getConst());
+
+          Teuchos::RCP<const ExportType> exporter =
+            Tpetra::createExport(locally_relevant_map.getConst(),
+                                 locally_owned_map.getConst());
+          dummy.doExport(*vector, *exporter, tpetra_operation);
+
+          *vector    = dummy;
```

Comment:
This is quite an expensive operation for `compress()`. Why is this necessary to make things work, given that that other vector types do not need to create such a temporary vector?

Reply:
It is not necessary to use a dummy vector here.

I learned this after I wrote that part of the class. I removed that dummy vector.

Again, thanks for pointing this out!

Reply:
I should have tested this before I answered; it seems like we need that dummy vector here.

It is not necessary in the Epetra-based vector class since a second vector, i.e., the `nonlocal_vector`, is already defined.